### PR TITLE
bin: check listen address and advertise address

### LINKED ISSUE #1332: Failed to start tikv-server cluster when just specify listen address
When starting multiple tikv with following command on different machine, only one can stay alive.

```
tikv-server --pd pd_addr -S raftkv -A 0.0.0.0:8000 -s store --log-file tikv.log --config etc/config-template.toml
```

All other instances will panic with following stacktrace:
```
thread 'main' panicked 'called `Result::unwrap()` on an `Err` value: Pd(Other(StringError("duplicated store address: id:8 address:\"0.0.0.0:8000\" , already registered by id:1 address:\"0.0.0.0:8000\" ")))' at "../src/libcore/result.rs:788"                                                                                                                                                                                       
stack backtrace:                                                                                                                                                                                                                              
   0:     0x7f41c4a48bfe - backtrace::backtrace::libunwind::trace                                                                                                                                                                             
                        at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54                                                                                                                
                         - backtrace::backtrace::trace<closure>                                                                                                                                                                               
                        at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70                                                                                                                      
   1:     0x7f41c4a49313 - backtrace::capture::{{impl}}::new                                                                                                                                                                                  
                        at /tikv/target/release/build/backtrace-3bc1ff360ebb00cb/out/capture.rs:79                                                                                                                                            
   2:     0x7f41c4627e20 - tikv::util::panic_hook::set_exit_hook::{{closure}}                                                                                                                                                                 
                        at /tikv/src/util/panic_hook.rs:84                                                                                                                                                                                    
   3:     0x7f41c4b3b307 - std::panicking::rust_panic_with_hook::h105c3d42fcd2fb5e                                                                                                                                                            
   4:     0x7f41c4b3b1f2 - std::panicking::begin_panic::hbf62ea4a5ff3f9de                                                                                                                                                                     
   5:     0x7f41c4b3b120 - std::panicking::begin_panic_fmt::h20f5943904e5791d                                                                                                                                                                 
   6:     0x7f41c4b3b0a1 - rust_begin_unwind                                                                                                                                                                                                  
   7:     0x7f41c4b7866c - core::panicking::panic_fmt::h19323e466869c656                                                                                                                                                                      
   8:     0x7f41c447b2a0 - {{inlined-root}}::unwrap_failed<tikv::server::errors::Error>                                                                                                                                                       
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/result.rs:29                                                                                                                     
   9:     0x7f41c44fbdea - {{inlined-root}}::unwrap<(),tikv::server::errors::Error>                                                                                                                                                           
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/result.rs:726                                                                                                                    
                         - tikv_server::build_raftkv                                                                                                                                                                                          
                        at /tikv/src/bin/tikv-server.rs:444                                                                                                                                                                                   
  10:     0x7f41c44ff27f - tikv_server::run_raft_server                                                                                                                                                                                       
                        at /tikv/src/bin/tikv-server.rs:546                                                                                                                                                                                   
  11:     0x7f41c4509277 - tikv_server::main                                                                                                                                                                                                  
                        at /tikv/src/bin/tikv-server.rs:666                                                                                                                                                                                   
  12:     0x7f41c4b3aff7 - std::panicking::try::call::h5df3ac2979db3c90                                                                                                                                                                       
  13:     0x7f41c4b43856 - __rust_maybe_catch_panic                                                                                                                                                                                           
  14:     0x7f41c4b3a199 - std::rt::lang_start::hfe9ab243c60ffb9b                                                                                                                                                                             
  15:     0x7f41c36fef44 - __libc_start_main                                                                                                                                                                                                  
  16:     0x7f41c4425e30 - <unknown>
```

### PR DESCRIPTION
**Pull request by overvenus**
If `addr` is `0.0.0.0` then `advertise-addr` is required.
Also `0.0.0.0` is not allowed in `advertise-addr`.

Close #1332 

PTAL @siddontang @BusyJay

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -180,6 +180,35 @@ fn check_system_config(config: &toml::Value) {
...
+
+    let all_addr = "0.0.0.0";
+    match addr.find(all_addr) {
+        Some(_) => {
+            info!("Start listening on {}... advertise-addr is required", addr);
+
+            let adv_addr = get_flag_string(matches, "advertise-addr")
+                .unwrap_or_else(|| get_toml_string(config, "server.advertise-addr", None));
+
+            if let Some(_) = adv_addr.find(all_addr) {
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: we should also check advertise-addr is 0.0.0.0 too, no matter addr is or not.

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -180,6 +180,29 @@ fn check_system_config(config: &toml::Value) {
...
+        .map_err(|e| exit_with_err(format!("{:?}", e)))
+        .unwrap();
+
+    let adv_addr = get_flag_string(matches, "advertise-addr")
+        .unwrap_or_else(|| get_toml_string(config, "server.advertise-addr", Some(addr.clone())));
+    util::config::check_addr(&adv_addr)
+        .map_err(|e| exit_with_err(format!("{:?}", e)))
+        .unwrap();
+
+    if let Some(_) = adv_addr.find("0.0.0.0") {
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: Oops, I will address this comment in other PR. :joy:

Comment (BusyJay) [inline src/bin/tikv-server.rs]: Is `*.*.*.0` a valid advertise address?

Comment (overvenus) [inline src/bin/tikv-server.rs]: I guess tikv will fail to bind such address.

Comment (BusyJay) [inline src/bin/tikv-server.rs]: What if I pass it to advertise address explicitly?

Comment (siddontang) [inline src/bin/tikv-server.rs]: There is an issue for this: https://github.com/pingcap/pd/issues/350

Comment (BusyJay) [inline src/bin/tikv-server.rs]: Should it be checked in TiKV too? So it can't setup if the address is invalid.

**Pull request merged**